include/ring_buffer.h

```diff
@@ -93,13 +97,16 @@ struct ring_buf {
  *
  * @code extern struct ring_buf <name>; @endcode
  *
+ * @note Buffer is allocated for one additional element which is used
+ * internally to distinguish between empty and full state.
+ *
  * @param name Name of the ring buffer.
  * @param size32 Size of ring buffer (in 32-bit words).
  */
 #define RING_BUF_ITEM_DECLARE_SIZE(name, size32) \
-	static u32_t _ring_buffer_data_##name[size32]; \
+	static u32_t _ring_buffer_data_##name[size32 + 1]; \
```

Comment:
can we just document that one element is always reserved, instead of changing how much memory get allocated?

Reply:
But from user perspective important thing is how many items i can fit in and not how much RAM it consumes. Ring buffer size may be tuned to maximal content size and then user will be disappointed that he needs to learn internals of ring buffer and add one more element.

In that case, i would change also the description:
`@param size32 Size of ring buffer (in 32-bit words).`
to:
`@param size32 Capacity of ring buffer (in 32-bit words).`

Did i convince you?

Reply:
> user will be disappointed that he needs to learn internals of ring buffer and add one more element. Did i convince you?

I for one not convinced. "One element of ringbuf should stay empty" is a common (external) API contract for ringbuf APIs. Moreover it was the API contract for Zephyr. You appear to change it, and changing something external is a "break". Let me give this -1 for more discussion.

Reply:
note that the only API breaker is for `RING_BUF_ITEM_DECLARE` since byte API is new. I can make `SYS_RING_BUF_DECLARE_SIZE` and `SYS_RING_BUF_DECLARE_POW2` backware compatible by doing `size32-1`.
As for new API I don't think it is a common API contract. I was involved in ring buffer changes and when working on shell uart which uses ring buffer I did not handle that case so from my experience i can say it is not natural to remember that ring buffer size does not mean capacity. 

Even [documentation](https://docs.zephyrproject.org/latest/kernel/other/ring_buffers.html) is misleading about that:
> A data buffer size, measured in 32-bit words or bytes. This governs the maximum amount of data (including metadata values) the ring buffer can hold.

So even initial author of ring buffer most likely forgot about 'less 1' case.

Reply:
> So even initial author of ring buffer most likely forgot about 'less 1' case during documentation phase.

It's subtle: "governs" doesn't mean "equals" ;-).

Reply:
From the docs:

> can be enqueued and dequeued from the ring buffer in chunks of up to 1020 bytes

A curious reader would definitely pay attention that it says 1020 bytes, not 1024 bytes.

Reply:
Gotcha:

https://docs.zephyrproject.org/latest/kernel/other/ring_buffers.html#internal-operation

> The ring buffer always maintains an empty 32-bit word (byte in bytes mode) in its data buffer to allow it to distinguish between empty and full states.

And what follows is an explanation why, for a low-language like C, "One element of ringbuf should stay empty" is the natural part of the API contract:

> If the size of the data buffer is a power of two, the ring buffer uses efficient masking operations instead of expensive modulo operations when enqueuing and dequeuing data items.

So, when you declare a "power of 2" buffer, makes sense to use numbers like 16, and not 15, when describing it. Heck, Zephyr's SYS_RING_BUF_DECLARE_POW2 actually even takes the power (logarithm) of 2, not the numeric length. How would you reconcile all that with the changes you propose? If we do it like that, there will be much more confusion, than what you try to solve. The current behavior is pretty much documented, though perhaps not explicitly enough. So, we should document it more explicitly.

Reply:
> A curious reader would definitely pay attention that it says 1020 bytes, not 1024 bytes.

This is something else: 4 byte header in case of 'packet mode'.

Regarding 'internal operation' section. Do you really expect user of ring buffer to read 'internal operations' section? One could as well change the implementation, add `item_cnt` field to the structure and remove need of spare element (in terms of RAM is equally good solution).

I really don't get this 'natural API contract'. Because we use low level language we must enforce user to understand modules internals? As a user i expect that when i do `RING_BUFFER_DEF(buf,16)` i can fit 16 elements there. I care much less how much RAM it takes and what kind of algorithm is used. I want my 16 elements to fit in that ring buffer.

Reply:
> This is something else: 4 byte header in case of 'packet mode'.

Ok, perhaps. (The docs are nowhere near detailed as perfect docs should be.)

> Regarding 'internal operation' section. Do you really expect user of ring buffer to read 'internal operations' section?

Generally, I expect users to real *all* related documentation, it's programming, and even embedded, low-level programming after all. And we already discussed what's wrong there - docs aren't good enough. For example what they title "internal operation" is actually external operation.

> One could as well change the implementation, add item_cnt field to the structure and remove need of spare element (in terms of RAM is equally good solution).

How a C programmer could say something like that? How can you compare 1 internally-reserved byte in a byte ringbuf with `size_t` typed item_cnt field - that's 8 bytes on 64-bit archs (have you heard what @andrewboie said (elsewhere) - they already work on 64-bit port. And that's 16 bytes on 128-bit arch (heard about 128-bit RISC-V yet?)

> I really don't get this 'natural API contract'.

That's why we discuss it. Because the alternative is worse - every Zephyr programmer wants to bring in some kind of new API, and every second wants to deprecate or change existing API. That doesn't scale. We need to break stuff - as an exception, under very good reasons, and hold ourselves tight in all other cases. (A colleague should help a colleague.)

> Because we use low level language we must enforce user to understand modules internals?

Again, that's part of external API definition (also known as "API contract").

> As a user i expect that when i do RING_BUFFER_DEF(buf,16) i can fit 16 elements there.

Wrong language then. I suggest to skip C++ and Rust and go directly to Python ;-).

> I care much less how much RAM it takes and what kind of algorithm is used.

You're in bad company then ;-), because all (most?) Zephyr programmers care about how much RAM it takes.

> I want my 16 elements to fit in that ring buffer.

Read the docs, note that you pass *size* to the constructor, not that capacity is described as being one less than size, so pass 16+1 to the constructor. (And be ready to various consequences, like 16-element byte ringbuf (just buf part) make take 32 bytes instead. Why? Allocation granularity for example.)